src/shared/completion/use-imperative-cancel-completion-mutation.js

```diff
@@ -0,0 +1,38 @@
+import { useQueryClient } from '@tanstack/react-query'
+import { useDataValueSetQueryKey } from '../use-data-value-set/index.js'
+import useSetFormCompletionMutationKey from './use-set-form-completion-mutation-key.js'
+
+export default function useImperativeCancelCompletionMutation() {
+    const queryClient = useQueryClient()
+    const mutationCache = queryClient.getMutationCache()
+    const mutationKey = useSetFormCompletionMutationKey()
+    const dataValueSetQueryKey = useDataValueSetQueryKey()
+
+    return () => {
+        const foundMutation = mutationCache.find({ mutationKey })
+
+        if (!foundMutation) {
+            return
+        }
+
+        // remove the mutation itself
+        mutationCache.remove(foundMutation)
+
+        // undo the optimistic update
+        // @TODO: Why does this cause a request to the data values that will
+        //   end up with a 404 due to being offline while the `onMutate`
+        //   callback in the `useSetFormCompletionMutation` does not?
+        const { completed } = foundMutation.options.variables
+
+        const previousDataValueSet =
+            queryClient.getQueryData(dataValueSetQueryKey)
+
+        queryClient.setQueryData(dataValueSetQueryKey, () => ({
```

Comment:
https://tanstack.com/query/v4/docs/reference/QueryClient#queryclientsetquerydata
This gets passed the old-data.

```js
queryClient.setQueryData(dataValueSetQueryKey, ({ completed: prevCompleted} ) => ({
```